File: app/src/main/java/com/hover/stax/home/MainActivity.java
package com.hover.stax.home;


import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.play.core.review.ReviewManager;
import com.google.android.play.core.review.ReviewManagerFactory;
import com.hover.sdk.actions.HoverAction;
import com.hover.stax.R;
import com.hover.stax.balances.BalanceAdapter;
import com.hover.stax.balances.BalancesViewModel;
import com.hover.stax.channels.Channel;
import com.hover.stax.databinding.ActivityMainBinding;
import com.hover.stax.hover.HoverSession;
import com.hover.stax.navigation.AbstractNavigationActivity;
import com.hover.stax.schedules.Schedule;
import com.hover.stax.settings.BiometricChecker;
import com.hover.stax.transactions.TransactionHistoryViewModel;
import com.hover.stax.utils.Constants;
import com.hover.stax.utils.DateUtils;
import com.hover.stax.utils.UIHelper;
import com.hover.stax.utils.Utils;

import java.util.List;

import timber.log.Timber;

public class MainActivity extends AbstractNavigationActivity implements
        BalancesViewModel.RunBalanceListener, BalanceAdapter.BalanceListener, BiometricChecker.AuthListener {

    final public static String TAG = "MainActivity";
    private BalancesViewModel balancesViewModel;

    private ActivityMainBinding binding;
    TransactionHistoryViewModel transactionHistoryViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        balancesViewModel = new ViewModelProvider(this).get(BalancesViewModel.class);
        transactionHistoryViewModel  = new ViewModelProvider(this).get(TransactionHistoryViewModel.class);
        balancesViewModel.setListener(this);
        balancesViewModel.getSelectedChannels().observe(this, channels -> Timber.i("Channels observer is necessary to make updates fire, but all logic is in viewmodel. %s", channels.size()));
        balancesViewModel.getToRun().observe(this, actions -> Timber.i("RunActions observer is necessary to make updates fire, but all logic is in viewmodel. %s", actions.size()));
        balancesViewModel.getRunFlag().observe(this, flag -> Timber.i("Flag observer is necessary to make updates fire, but all logic is in viewmodel. %s", flag));
        balancesViewModel.getActions().observe(this, actions -> Timber.i("Actions observer is necessary to make updates fire, but all logic is in viewmodel. %s", actions.size()));

        checkForRequest(getIntent());
        checkForFragmentDirection(getIntent());
        checkForDeepLinking();
        observeForAppReview();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        checkForRequest(intent);
    }

    @Override
    protected void onResume() {
        super.onResume();

        setUpNav();
    }

    private void checkForDeepLinking() {
        Intent intent = getIntent();
        if(intent.getAction()!=null && intent.getAction().equals(Intent.ACTION_VIEW) && intent.getData() != null) {
            String deepLinkRoute = intent.getData().toString();
            if(deepLinkRoute.contains(getString(R.string.deeplink_sendmoney))) {
                navigateToTransferActivity(HoverAction.P2P, false, intent, this);
            }
            else if(deepLinkRoute.contains(getString(R.string.deeplink_airtime))) {
                navigateToTransferActivity(HoverAction.AIRTIME, false, intent, this);
            }
            else if(deepLinkRoute.contains(getString(R.string.deeplink_linkaccount))) {
                navigateToChannelsListFragment(getNavController(), true);
            }
            else if(deepLinkRoute.contains(getString(R.string.deeplink_balance)) || deepLinkRoute.contains(getString(R.string.deeplink_history))) {
                navigateToBalanceFragment(getNavController());
            }
            else if(deepLinkRoute.contains(getString(R.string.deeplink_settings))) {
                navigateToSettingsFragment(getNavController());
            }
            else if(deepLinkRoute.contains(getString(R.string.deeplink_reviews))) {
                Timber.i("Currently at reviews 1");
                launchStaxReview();
            }
        }
    }
    private void observeForAppReview() {
        transactionHistoryViewModel.showAppReviewLiveData().observe(this, status -> {
            if(status) { launchRatingAndReviewDialog(); }
        });
    }

    private void launchStaxReview() {
        Utils.logAnalyticsEvent(getString(R.string.visited_rating_review_screen), this);
        if(Utils.getBoolean(Constants.APP_RATED_NATIVELY, this)) openStaxPlaystorePage();
        else launchRatingAndReviewDialog();
    }

    private void launchRatingAndReviewDialog() {
        Timber.i("Currently at reviews native");
        ReviewManager reviewManager = ReviewManagerFactory.create(this);
        reviewManager.requestReviewFlow().addOnCompleteListener(task -> {
            Timber.i("Currently at reviews native already completed");
            if(task.isSuccessful()){
                Timber.i("Currently at reviews native already succeeded");
                reviewManager.launchReviewFlow(MainActivity.this, task.getResult()).addOnCompleteListener(
                        task1 -> Utils.saveBoolean(Constants.APP_RATED_NATIVELY, true, MainActivity.this));
            }
        });
    }

    private void openStaxPlaystorePage() {
        Timber.i("Currently at reviews for external store page");
        Uri link = Uri.parse(getString(R.string.stax_market_playstore_link));
        Intent goToMarket = new Intent(Intent.ACTION_VIEW, link);
        goToMarket.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY | Intent.FLAG_ACTIVITY_NEW_DOCUMENT | Intent.FLAG_ACTIVITY_MULTIPLE_TASK);
        try {
            startActivity(goToMarket);
        } catch (ActivityNotFoundException e) {
            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(getString(R.string.stax_url_playstore_review_link))));
        }
    }

    private void checkForRequest(Intent intent) {
        if (intent.hasExtra(Constants.REQUEST_LINK))
            navigateToTransferActivity(HoverAction.P2P, true, intent, this);
    }

    private void checkForFragmentDirection(Intent intent) {
        if (intent.hasExtra(Constants.FRAGMENT_DIRECT)) {
            int toWhere = intent.getExtras().getInt(Constants.FRAGMENT_DIRECT, 0);
            Timber.e("Dest %s", getNavConst(toWhere));

            checkPermissionsAndNavigate(toWhere);
        }
    }

    @Override
    public void onTapDetail(int channel_id) {
        if(channel_id == Channel.DUMMY) navigateToChannelsListFragment(getNavController(), true);
        else navigateToChannelDetailsFragment(channel_id, getNavController());
    }

    @Override
    public void onTapRefresh(int channel_id) {
        if(channel_id == Channel.DUMMY) navigateToChannelsListFragment(getNavController(), false);
        else{
            Utils.logAnalyticsEvent(getString(R.string.refresh_balance_single), this);
            balancesViewModel.setRunning(channel_id);
        }

    }

    @Override
    public void startRun(HoverAction a, int i) {
        run(a, i);
    }

    private void run(HoverAction action, int index) {
        Timber.e("running index: %s", index);

        if (balancesViewModel.getChannel(action.channel_id) != null) {
            HoverSession.Builder hsb = new HoverSession.Builder(action, balancesViewModel.getChannel(action.channel_id), MainActivity.this, index);
            if (index + 1 < balancesViewModel.getSelectedChannels().getValue().size())
                hsb.finalScreenTime(0);
            hsb.run();
        } else { // Fix for auth issue on OnePlus 6
            new Handler(Looper.getMainLooper()).post(() -> {
                balancesViewModel.getSelectedChannels().observe(MainActivity.this, new Observer<List<Channel>>() {
                    @Override
                    public void onChanged(List<Channel> channels) {
                        if (channels != null && balancesViewModel.getChannel(channels, action.channel_id) != null) {
                            run(action, 0);
                            balancesViewModel.getSelectedChannels().removeObserver(this);
                        }
                    }
                });
            });
        }
    }

    @Override
    public void onAuthError(String error) {
        Timber.e("error: %s", error);
    }

    @Override
    public void onAuthSuccess(HoverAction act) {
        run(act, 0);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case Constants.TRANSFER_REQUEST:
                if (data != null) {
                    onProbableHoverCall(data);
                }
                break;
            case Constants.REQUEST_REQUEST:
                if (resultCode == RESULT_OK) {
                    onRequest(data);
                }
                break;
            default: // requestCode < Constants.BALANCE_MAX // Balance call
                balancesViewModel.setRan(requestCode);
                if (resultCode == RESULT_OK && data != null && data.getAction() != null)
                    onProbableHoverCall(data);
        }
    }

    private void onProbableHoverCall(Intent data) {
        if (data.getAction() != null && data.getAction().equals(Constants.SCHEDULED)) {
            showMessage(getString(R.string.toast_confirm_schedule, DateUtils.humanFriendlyDate(data.getLongExtra(Schedule.DATE_KEY, 0))));
        } else {
            Utils.logAnalyticsEvent(getString(R.string.finish_load_screen), this);
            new ViewModelProvider(this).get(TransactionHistoryViewModel.class).saveTransaction(data, this);
        }
    }


    private void onRequest(Intent data) {
        if (data.getAction().equals(Constants.SCHEDULED))
            showMessage(getString(R.string.toast_request_scheduled, DateUtils.humanFriendlyDate(data.getLongExtra(Schedule.DATE_KEY, 0))));
        else
            showMessage(getString(R.string.toast_confirm_request));
    }

    private void showMessage(String str) {
        UIHelper.flashMessage(this, findViewById(R.id.fab), str);
    }
}

File: app/src/main/java/com/hover/stax/home/MainActivity.kt
/*
 * Copyright 2022 Stax
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hover.stax.home

import android.content.Intent
import android.os.Bundle
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.NavDirections
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.hover.sdk.actions.HoverAction
import com.hover.sdk.api.Hover
import com.hover.sdk.permissions.PermissionHelper
import com.hover.stax.FRAGMENT_DIRECT
import com.hover.stax.MainNavigationDirections
import com.hover.stax.R
import com.hover.stax.databinding.ActivityMainBinding
import com.hover.stax.login.AbstractGoogleAuthActivity
import com.hover.stax.login.LoginViewModel
import com.hover.stax.login.StaxGoogleLoginInterface
import com.hover.stax.notifications.PushNotificationTopicsInterface
import com.hover.stax.presentation.bounties.BountyApplicationFragmentDirections
import com.hover.stax.presentation.financial_tips.FinancialTipsFragment
import com.hover.stax.requests.NewRequestViewModel
import com.hover.stax.requests.REQUEST_LINK
import com.hover.stax.requests.RequestSenderInterface
import com.hover.stax.requests.SMS
import com.hover.stax.settings.BiometricChecker
import com.hover.stax.transactions.TransactionHistoryViewModel
import com.hover.stax.transfers.TransferViewModel
import com.hover.stax.utils.UIHelper
import com.hover.stax.views.StaxDialog
import dagger.hilt.android.AndroidEntryPoint
import timber.log.Timber

@AndroidEntryPoint
class MainActivity : AbstractGoogleAuthActivity(), BiometricChecker.AuthListener, PushNotificationTopicsInterface, RequestSenderInterface {

    private lateinit var viewModelFactory: ViewModelProvider.Factory

    private lateinit var loginViewModel: LoginViewModel
    private lateinit var staxGoogleLoginInterface: StaxGoogleLoginInterface

    override fun provideLoginViewModel(): LoginViewModel {
        val loginViewModel: LoginViewModel by viewModels()
        return loginViewModel
    }
    lateinit var navHelper: NavHelper

    private val transferViewModel: TransferViewModel by viewModels()
    private val requestViewModel: NewRequestViewModel by viewModels()
    private val historyViewModel: TransactionHistoryViewModel by viewModels()

    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        loginViewModel = provideLoginViewModel()
        staxGoogleLoginInterface = this
        viewModelFactory = ViewModelProvider.AndroidViewModelFactory.getInstance(application)

        navHelper = NavHelper(this)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        navHelper.setUpNav()
        initGoogleAuth()
        setLoginObserver()
        initFromIntent()
        checkForRequest(intent)
        checkForFragmentDirection(intent)
        observeForAppReview()
        setGoogleLoginInterface(this)
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        checkForRequest(intent!!)
        checkForFragmentDirection(intent)
    }

    override fun onResume() {
        super.onResume()
        navHelper.setUpNav()
    }

    fun checkPermissionsAndNavigate(navDirections: NavDirections) {
        navHelper.checkPermissionsAndNavigate(navDirections)
    }

    private fun observeForAppReview() = historyViewModel.showAppReviewLiveData().observe(this@MainActivity) {
        if (it) StaxAppReview.launchStaxReview(this@MainActivity)
    }

    private fun checkForRequest(intent: Intent) {
        if (intent.hasExtra(REQUEST_LINK)) {
            createFromRequest(intent.getStringExtra(REQUEST_LINK)!!)
        }
    }

    private fun checkForFragmentDirection(intent: Intent) {
        if (intent.hasExtra(FRAGMENT_DIRECT)) {
            val toWhere = intent.extras!!.getInt(FRAGMENT_DIRECT, 0)

            if (toWhere == NAV_EMAIL_CLIENT)
                com.hover.stax.utils.Utils.openEmail(getString(R.string.stax_emailing_subject, Hover.getDeviceId(this)), this)
            else
                navHelper.checkPermissionsAndNavigate(toWhere)
        }
    }
    private fun initGoogleAuth() {
        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestIdToken(getString(R.string.google_server_client_id)).requestEmail().build()
        loginViewModel.signInClient = GoogleSignIn.getClient(this, gso)
    }

    private fun setLoginObserver() = with(loginViewModel) {
        error.observe(this@MainActivity) {
            it?.let { staxGoogleLoginInterface.googleLoginFailed() }
        }

        googleUser.observe(this@MainActivity) {
            it?.let { staxGoogleLoginInterface.googleLoginSuccessful() }
        }
    }

    fun signIn() = loginForResult.launch(loginViewModel.signInClient.signInIntent)

    private val loginForResult =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == RESULT_OK) {
                loginViewModel.signIntoGoogle(result.data)
            } else {
                Timber.e("Google sign in failed")
                staxGoogleLoginInterface.googleLoginFailed()
            }
        }
    private fun initFromIntent() {
        when {
            intent.hasExtra(REQUEST_LINK) -> createFromRequest(intent.getStringExtra(REQUEST_LINK)!!)
            intent.hasExtra(FinancialTipsFragment.TIP_ID) -> navHelper.navigateWellness(intent.getStringExtra(FinancialTipsFragment.TIP_ID)!!)

            else -> com.hover.stax.utils.AnalyticsUtil.logAnalyticsEvent(getString(R.string.visit_screen, intent.action), this)
        }
    }

    private fun createFromRequest(link: String) {
        navHelper.checkPermissionsAndNavigate(MainNavigationDirections.actionGlobalTransferFragment(HoverAction.P2P))
        addLoadingDialog()
        transferViewModel.load(link)
        com.hover.stax.utils.AnalyticsUtil.logAnalyticsEvent(getString(R.string.clicked_request_link), this)
    }

    private fun addLoadingDialog() {
        val alertDialog = StaxDialog(this).setDialogMessage(R.string.loading_link_dialoghead).showIt()
        transferViewModel.isLoading.observe(this@MainActivity) { if (!it) alertDialog?.dismiss() }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == SMS && PermissionHelper(this).permissionsGranted(grantResults)) {
            com.hover.stax.utils.AnalyticsUtil.logAnalyticsEvent(getString(R.string.perms_sms_granted), this)
            sendSms(requestViewModel, this)
        } else if (requestCode == SMS) {
            com.hover.stax.utils.AnalyticsUtil.logAnalyticsEvent(getString(R.string.perms_sms_denied), this)
            UIHelper.flashAndReportMessage(this, getString(R.string.toast_error_smsperm))
        }
    }

    override fun googleLoginSuccessful() {
        if (loginViewModel.staxUser.value?.isMapper == true) BountyApplicationFragmentDirections.actionBountyApplicationFragmentToBountyListFragment()
    }

    override fun onAuthError(error: String) {
        Timber.e("Error : $error")
    }

    override fun onAuthSuccess(action: HoverAction?) {
        Timber.v("Auth success on action: ${action?.public_id}")
    }
}